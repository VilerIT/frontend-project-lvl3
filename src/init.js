import axios from 'axios';

import initView from './view.js';
import validateLink from './validateLink.js';
import parseRSS from './parseRSS.js';

const routes = {
  allorigins: (url) => `https://hexlet-allorigins.herokuapp.com/get?url=${encodeURIComponent(url)}`,
};

export default () => {
  const state = {
    rssForm: {
      state: 'filling',
      valid: false,
      error: null,
      isSuccess: false,
    },
    feeds: [],
    posts: [],
    uiState: {
      activeFeedId: null,
    },
  };

  const watchedState = initView(state);

  const form = document.querySelector('.rss-form');
  const input = document.querySelector('.form-control');

  form.addEventListener('submit', (e) => {
    e.preventDefault();

    const formData = new FormData(e.target);
    const link = formData.get('url');

    const error = validateLink(link, watchedState.feeds);
    watchedState.rssForm.error = error;

    if (error) {
      watchedState.rssForm.isSuccess = false;
    }

    watchedState.rssForm.valid = !error;

    if (watchedState.rssForm.valid) {
      watchedState.rssForm.state = 'pending';

      const feedId = watchedState.feeds.length + 1;

      axios.get(routes.allorigins(link))
        .then((response) => parseRSS(response.data.contents))
        .then((rss) => {
          const title = rss.querySelector('title').textContent;
          const desc = rss.querySelector('description').textContent;

          const posts = rss.querySelectorAll('item')
            .forEach((post) => {
              const postTitle = post.querySelector('title').textContent;
              const postDesc = post.querySelector('description').textContent;
              const postLink = post.querySelector('link').textContent;

              const postId = state.posts.length + 1;

              const data = {
                id: postId, feedId, title: postTitle, desc: postDesc, url: postLink,
              };

              state.posts.push(data);
            });

          const newFeed = {
            id: feedId, title, desc, url: link, posts,
          };

          watchedState.feeds.push(newFeed);
          watchedState.uiState.activeFeedId = feedId;
          watchedState.rssForm.isSuccess = true;

          form.reset();
          input.focus();
        })
        .catch(() => {
          watchedState.rssForm.isSuccess = false;
          watchedState.rssForm.error = 'Something went wrong';
        })
        .finally(() => {
          watchedState.rssForm.state = 'filling';
        });
    }
  });
};
