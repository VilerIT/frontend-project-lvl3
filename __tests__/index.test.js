import axios from 'axios';
import adapter from 'axios/lib/adapters/http';
import { fileURLToPath } from 'url';
import path, { dirname } from 'path';
import { promises as fs } from 'fs';
import { screen } from '@testing-library/dom/dist/screen.js';
import { fireEvent } from '@testing-library/dom/dist/events.js';
import { waitFor } from '@testing-library/dom/dist/wait-for.js';
import nock from 'nock';
import '@testing-library/jest-dom';

import init from '../src/init.js';

nock.disableNetConnect();

axios.defaults.adapter = adapter;

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const getFixturePath = (filename) => path.join(__dirname, '..', '__fixtures__', filename);

const readFixture = (filename) => fs.readFile(getFixturePath(filename), 'utf-8');

beforeEach(async () => {
  const html = await readFixture('index.html');
  document.body.innerHTML = html;

  await init();
});

test('init', () => {
  expect(document.body.innerHTML).toMatchSnapshot();
});

test('add rss feed', async () => {
  const requestURL = 'https://google.com';
  const fakeRSS = await readFixture('rss1.xml');

  nock('https://hexlet-allorigins.herokuapp.com')
    .get(`/get?url=${encodeURIComponent(requestURL)}&disableCache=true`)
    .reply(200, { contents: fakeRSS });

  fireEvent.input(screen.getByTestId('input'), { target: { value: requestURL } });
  fireEvent.submit(screen.getByTestId('form'));

  const success = await screen.findByText('RSS successfully loaded');
  expect(success).toBeInTheDocument();

  expect(document.body.innerHTML).toMatchSnapshot();
});

test('open/close modal', async () => {
  const requestURL = 'https://google.com';
  const fakeRSS = await readFixture('rss1.xml');

  nock('https://hexlet-allorigins.herokuapp.com')
    .get(`/get?url=${encodeURIComponent(requestURL)}&disableCache=true`)
    .reply(200, { contents: fakeRSS });

  fireEvent.input(screen.getByTestId('input'), { target: { value: requestURL } });
  fireEvent.submit(screen.getByTestId('form'));

  const viewButtons = await screen.findAllByText('View');
  fireEvent.click(viewButtons[0]);

  const modalTitle = await screen.findByTestId('modal-title');
  const modalBody = await screen.findByTestId('modal-body');

  expect(modalTitle).toHaveTextContent('Woohoo');
  expect(modalBody).toHaveTextContent('1111');

  fireEvent.click(screen.getByTestId('close-modal'));

  await waitFor(() => {
    expect(screen.getByTestId('modal')).not.toHaveClass('show');
  });

  expect(document.body.innerHTML).toMatchSnapshot();
});
