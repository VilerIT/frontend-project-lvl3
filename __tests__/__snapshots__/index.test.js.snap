// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add rss feed 1`] = `
"<div class=\\"modal fade\\" id=\\"modal\\" data-testid=\\"modal\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-labelledby=\\"modal\\" aria-hidden=\\"true\\">
  <div class=\\"modal-dialog\\" role=\\"document\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <h5 class=\\"modal-title\\" data-testid=\\"modal-title\\"></h5>
        <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\">×</span></button>
      </div>
      <div class=\\"modal-body\\" data-testid=\\"modal-body\\"></div>
      <div class=\\"modal-footer\\"><a class=\\"btn btn-primary full-article\\" href=\\"#\\" role=\\"button\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Read the article</a><button type=\\"button\\" data-testid=\\"close-modal\\" class=\\"btn btn-secondary\\" data-dismiss=\\"modal\\">Close</button></div>
    </div>
  </div>
</div>
<main class=\\"flex-grow-1\\">
  <section class=\\"container-fluid bg-dark p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
        <h1 class=\\"display-3\\">RSS Aggregator</h1>
        <p class=\\"lead\\">Start reading RSS today! It's easy, it's beautiful.</p>
        <div class=\\"btn-group btn-group-toggle\\" data-toggle=\\"buttons\\" style=\\"padding-bottom: 15px;\\">
          <label class=\\"btn btn-light active\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"en\\" checked=\\"\\"> English
          </label>
          <label class=\\"btn btn-outline-light\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"ru\\"> Русский
          </label>
        </div>
        <form action=\\"\\" class=\\"rss-form\\" data-testid=\\"form\\">
          <div class=\\"form-row\\">
            <div class=\\"col\\"><input autofocus=\\"\\" required=\\"\\" name=\\"url\\" data-testid=\\"input\\" aria-label=\\"url\\" class=\\"form-control form-control-lg w-100\\" placeholder=\\"RSS link\\"></div>
            <div class=\\"col-auto\\"><button type=\\"submit\\" aria-label=\\"add\\" class=\\"btn btn-lg btn-primary px-sm-5\\">Add</button></div>
          </div>
        </form>
        <p class=\\"text-muted my-1\\">Example link: https://www.theguardian.com/international/rss</p>
        <div class=\\"feedback text-success\\">RSS successfully loaded</div>
      </div>
    </div>
  </section>
  <section class=\\"container-fluid p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\"><h2>Feeds</h2><ul class=\\"list-group mb-5\\"><li class=\\"list-group-item\\" data-id=\\"1\\">
    <h3>Test RSS 1</h3>
    <p>Test desc 1</p>
    </li></ul></div>
    </div>
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto posts\\"><h2>Posts</h2><ul class=\\"list-group\\"><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"2\\">
    <a href=\\"https://google.com\\" class=\\"font-weight-bold\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      Woohoo
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"3\\">
    <a href=\\"https://example.com\\" class=\\"font-weight-bold\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      Hey
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"4\\">
    <a href=\\"https://yahoo.com\\" class=\\"font-weight-bold\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      LOL
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li></ul></div>
    </div>
  </section>
</main>
<footer class=\\"d-flex justify-content-around py-3\\">
  <span>© 2021 Developed by <a href=\\"https://github.com/VilerIT\\">VilerIT</a> (Zaur Radjabov)</span>
</footer>
"
`;

exports[`init 1`] = `
"<div class=\\"modal fade\\" id=\\"modal\\" data-testid=\\"modal\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-labelledby=\\"modal\\" aria-hidden=\\"true\\">
  <div class=\\"modal-dialog\\" role=\\"document\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <h5 class=\\"modal-title\\" data-testid=\\"modal-title\\"></h5>
        <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\">×</span></button>
      </div>
      <div class=\\"modal-body\\" data-testid=\\"modal-body\\"></div>
      <div class=\\"modal-footer\\"><a class=\\"btn btn-primary full-article\\" href=\\"#\\" role=\\"button\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Read the article </a><button type=\\"button\\" data-testid=\\"close-modal\\" class=\\"btn btn-secondary\\" data-dismiss=\\"modal\\">Close</button></div>
    </div>
  </div>
</div>
<main class=\\"flex-grow-1\\">
  <section class=\\"container-fluid bg-dark p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
        <h1 class=\\"display-3\\">RSS Aggregator</h1>
        <p class=\\"lead\\">Start reading RSS today! It's easy, it's beautiful.</p>
        <div class=\\"btn-group btn-group-toggle\\" data-toggle=\\"buttons\\" style=\\"padding-bottom: 15px;\\">
          <label class=\\"btn btn-light active\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"en\\" checked=\\"\\"> English
          </label>
          <label class=\\"btn btn-outline-light\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"ru\\"> Русский
          </label>
        </div>
        <form action=\\"\\" class=\\"rss-form\\" data-testid=\\"form\\">
          <div class=\\"form-row\\">
            <div class=\\"col\\"><input autofocus=\\"\\" required=\\"\\" name=\\"url\\" data-testid=\\"input\\" aria-label=\\"url\\" class=\\"form-control form-control-lg w-100\\" placeholder=\\"RSS link\\"></div>
            <div class=\\"col-auto\\"><button type=\\"submit\\" aria-label=\\"add\\" class=\\"btn btn-lg btn-primary px-sm-5\\">Add</button></div>
          </div>
        </form>
        <p class=\\"text-muted my-1\\">Example link: https://www.theguardian.com/international/rss</p>
        <div class=\\"feedback\\"></div>
      </div>
    </div>
  </section>
  <section class=\\"container-fluid p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\"></div>
    </div>
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto posts\\"></div>
    </div>
  </section>
</main>
<footer class=\\"d-flex justify-content-around py-3\\">
  <span>© 2021 Developed by <a href=\\"https://github.com/VilerIT\\">VilerIT</a> (Zaur Radjabov)</span>
</footer>
"
`;

exports[`open/close modal 1`] = `
"<div class=\\"modal fade\\" id=\\"modal\\" data-testid=\\"modal\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-labelledby=\\"modal\\" aria-hidden=\\"true\\" style=\\"padding-right: 0px;\\">
  <div class=\\"modal-dialog\\" role=\\"document\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <h5 class=\\"modal-title\\" data-testid=\\"modal-title\\">Woohoo</h5>
        <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\">×</span></button>
      </div>
      <div class=\\"modal-body\\" data-testid=\\"modal-body\\">1111</div>
      <div class=\\"modal-footer\\"><a class=\\"btn btn-primary full-article\\" href=\\"https://google.com\\" role=\\"button\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Read the article</a><button type=\\"button\\" data-testid=\\"close-modal\\" class=\\"btn btn-secondary\\" data-dismiss=\\"modal\\">Close</button></div>
    </div>
  </div>
</div>
<main class=\\"flex-grow-1\\">
  <section class=\\"container-fluid bg-dark p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
        <h1 class=\\"display-3\\">RSS Aggregator</h1>
        <p class=\\"lead\\">Start reading RSS today! It's easy, it's beautiful.</p>
        <div class=\\"btn-group btn-group-toggle\\" data-toggle=\\"buttons\\" style=\\"padding-bottom: 15px;\\">
          <label class=\\"btn btn-light active\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"en\\" checked=\\"\\"> English
          </label>
          <label class=\\"btn btn-outline-light\\">
            <input type=\\"radio\\" autocomplete=\\"off\\" data-lang=\\"ru\\"> Русский
          </label>
        </div>
        <form action=\\"\\" class=\\"rss-form\\" data-testid=\\"form\\">
          <div class=\\"form-row\\">
            <div class=\\"col\\"><input autofocus=\\"\\" required=\\"\\" name=\\"url\\" data-testid=\\"input\\" aria-label=\\"url\\" class=\\"form-control form-control-lg w-100\\" placeholder=\\"RSS link\\"></div>
            <div class=\\"col-auto\\"><button type=\\"submit\\" aria-label=\\"add\\" class=\\"btn btn-lg btn-primary px-sm-5\\">Add</button></div>
          </div>
        </form>
        <p class=\\"text-muted my-1\\">Example link: https://www.theguardian.com/international/rss</p>
        <div class=\\"feedback text-success\\">RSS successfully loaded</div>
      </div>
    </div>
  </section>
  <section class=\\"container-fluid p-5\\">
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\"><h2>Feeds</h2><ul class=\\"list-group mb-5\\"><li class=\\"list-group-item\\" data-id=\\"5\\">
    <h3>Test RSS 1</h3>
    <p>Test desc 1</p>
    </li></ul></div>
    </div>
    <div class=\\"row\\">
      <div class=\\"col-md-10 col-lg-8 mx-auto posts\\"><h2>Posts</h2><ul class=\\"list-group\\"><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"6\\">
    <a href=\\"https://google.com\\" class=\\"font-weight-normal\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      Woohoo
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"7\\">
    <a href=\\"https://example.com\\" class=\\"font-weight-bold\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      Hey
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li><li class=\\"list-group-item d-flex justify-content-between align-items-start\\" data-id=\\"8\\">
    <a href=\\"https://yahoo.com\\" class=\\"font-weight-bold\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
      LOL
    </a>
    <button type=\\"button\\" class=\\"btn btn-primary btn-sm\\">View</button>
    </li></ul></div>
    </div>
  </section>
</main>
<footer class=\\"d-flex justify-content-around py-3\\">
  <span>© 2021 Developed by <a href=\\"https://github.com/VilerIT\\">VilerIT</a> (Zaur Radjabov)</span>
</footer>
<div class=\\"modal-backdrop fade show\\"></div>"
`;
